pkg/kmodule/kmodule_linux.go

```diff
@@ -237,31 +216,72 @@ func loadDeps(path string, m depMap, opts ProbeOpts) error {
 	}
 
 	// done with dependencies, load module
-	if err := loadModule(path, "", opts); err != nil {
+	if err := LoadModule(path, "", 0, opts); err != nil {
 		return err
 	}
 	m[path].state = loaded
 
 	return nil
 }
 
-func loadModule(path, modParams string, opts ProbeOpts) error {
+// loadModule loads a file as a module. Because of all the various hacks -- er,
+// improvements, made to module loading over 20 years, this gets messy.
+// o finit_module is optional and only takes an fd
+// o you can only use flags with finit_module
+// o some distros have compressed modules
+//   - can you uncompress those to a file?
+//     what if /tmp is not there for some reason? What if the module you are
+//     loading is tmpfs and you can't make temp files until you modprobe tmpfs?
+//     [this can happen]
+//   - Can you write the uncompressed data to a pipe and pass that fd?
+//     You may not be able to write all the data to a pipe. 4k is a common size.
+//     What if the module loader stats the pipe? It won't get the right answer.
+//     What if the writer gets behind and kernel reads 0 bytes for some reason and assumes EOF?
+//     What if a writer messes up due to bad compression and you block the kernel?
+// It's a mess.
+//
+// With luck, this function covers all the cases.
+// I put it back in one function as I found it a bit more readable straight line --
+// it's less than a page.
+// Note that, at present, we don't even set flags. Arguably, using FinitModule
+// is a waste of time.
+func LoadModule(path, modParams string, flags int, opts ProbeOpts) error {
 	if opts.DryRunCB != nil {
 		opts.DryRunCB(path)
 		return nil
 	}
 
 	f, err := os.Open(path)
-	if err != nil {
-		return err
-	}
 	defer f.Close()
+	if err == nil {
+		// Use e instead of err to ensure visitors from the future don't see err
+		// and think the := is a typo.
+		if e := unix.FinitModule(int(f.Fd()), modParams, flags); e != unix.ENOSYS {
+			return e
+		}
+	}
+	if flags != 0 {
+		return fmt.Errorf("Can not have non-zero flags (%#x) with init_module", flags)
+	}
 
-	if err := FileInit(f, modParams, 0); err != nil && err != unix.EEXIST {
-		return err
+	// we're here; if the file open failed, try an xz file.
+	var r = io.Reader(f)
```

Comment:
If `f == nil`, won't this be an issue? Maybe put `r = io.Reader(f)` in an else clause.